### PR TITLE
#525 List search: added SearchBar, added new filter to inventory

### PR DESCRIPTION
**Pull request by az1plo**
**What this PR does**

- Added SearchBar component for filtering lists.
- Integrated SearchBar into the following sections:
   - Inventory
   - Shop
   - Workshop
   - Cows
   - Cellar
- Implemented a new filter logic for the inventory to enhance usability.

**How this change can be validated**

- Navigate to each section listed above and use the search bar to filter items.
- Verify that the displayed items match the search query.

### Questions or concerns about this change

- Are there sections where the SearchBar could be removed or added?

### Additional information

Screenshot 2024-11-25 113651
Screenshot 2024-11-25 113712
Screenshot 2024-11-25 113916
Screenshot 2024-11-25 113951
Screenshot 2024-11-25 114026
Screenshot 2024-11-25 113846

### CODE REVIEW

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
Thank you for the amazing work so far @az1plo! This is working really well. There are a number of UI consistency and performance issues I'd like to see addressed before merging this. Hopefully my comments aren't too onerous to deal with. Feel free to ask any followup questions here or in Discord!

File: src/components/FermentationRecipeList/FermentationRecipeList.test.js
Comment (luke-createdbyhumans) [inline src/components/FermentationRecipeList/FermentationRecipeList.test.js:79]: if possible, we should use `userEvent.type` instead of `fireEvent`. note: it requires an `await`. docs here https://github.com/testing-library/user-event (i'm only going to leave this comment here, but i see `fireEvent` in some other tests that i think could be updated as well)

**Review by luke-createdbyhumans — COMMENTED** [review]
i didn't have time to fully review yet, but one thing did stand out to me and i'm also seeing that when the search input is not in the left column (like when searching the kitchen or the shop) it needs some more horizontal padding to line up with the other items in that view. 

it's great to see this PR though! i'll look at it more closely when i can

File: src/components/FermentationRecipeList/FermentationRecipeList.test.js
Comment (lstebner) [inline src/components/FermentationRecipeList/FermentationRecipeList.test.js:79]: if possible, we should use userEvent.type instead of fireEvent. note: it requires an await. docs here https://github.com/testing-library/user-event (i'm only going to leave this comment here, but i see fireEvent in some other tests that i think could be updated as well)

**Review by lstebner — COMMENTED** [review]
sorry i'm re-posting my comments from my personal account because i was logged in to the wrong account at first

i didn't have time to fully review yet, but one thing did stand out to me and i'm also seeing that when the search input is not in the left column (like when searching the kitchen or the shop) it needs some more horizontal padding to line up with the other items in that view.

it's great to see this PR though! i'll look at it more closely when i can

File: src/components/SearchBar/SearchBar.test.js
Comment (jeremyckahn) [inline src/components/SearchBar/SearchBar.test.js:30]: Let's translate this comment to English as well. 🙂

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
Great work @az1plo! There's just a few more things to tighten up before this can be merged. I think we're almost there!

File: package.json
Comment (jeremyckahn) [inline package.json:76]: Let remove the inclusion of Jest here. This project [uses Vitest](https://github.com/jeremyckahn/farmhand/blob/6419076acb608c99baf357e8b0851ee4ae7eb1b0/package.json#L93), so this package isn't necessary.

File: src/components/Inventory/Inventory.js
Comment (jeremyckahn) [inline src/components/Inventory/Inventory.js:16]: :heart:

File: src/components/Inventory/Inventory.test.js
Comment (jeremyckahn) [inline src/components/Inventory/Inventory.test.js:21]: Let's remove this to keep test output clean.

File: src/components/Inventory/Inventory.test.js
Comment (jeremyckahn) [inline src/components/Inventory/Inventory.test.js:1]: It looks like there are two test failures happening in this file. These will need to be resolved before this can be merged.

---

This would have been caught by the CI system, but it appears that the tests aren't being run for this PR. It's not clear why that is, as all the configuration and GitHub Actions setup seems correct. Since GitHub doesn't appear to be cooperating, you'll have to run the tests manually, locally. :confused:

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
Thanks for the latest changes @az1plo! There's a few small test-related issues to iron out, but the application code is looking good. 🙂

File: src/components/Inventory/Inventory.test.js
Comment (jeremyckahn) [inline src/components/Inventory/Inventory.test.js:55]: This test seems valuable. Would you mind reverting this removal (and make any necessary updates to make it pass)?

File: package.json
Comment (jeremyckahn) [reply package.json:76]: It looks like this still needs to be reverted out of this PR.

**Review by jeremyckahn — CHANGES_REQUESTED** [review]
I pushed https://github.com/jeremyckahn/farmhand/pull/526/commits/fdef440283f348aaa7b0d14ae3374a47c9d983e8 to address the test failures, so we should be good there.

Just a few more test-related items to address and this will be good to merge!

File: src/components/Inventory/Inventory.test.js
Comment (jeremyckahn) [inline src/components/Inventory/Inventory.test.js:77]: Awesome! Thanks for adding this. 🙂

**Review by jeremyckahn — APPROVED** [review]
Great work @az1plo! This looks and works great. Approved! :clap:

I'm going to merge this now, but I'm going to hold off on releasing it to Production because I'd like to get #527 merged ahead of the next release. Releasing the changes together will result in less update churn for players. I should have a fix in for that soon and will announce the release on Discord.

Thanks again for the amazing contribution! :raised_hands:

**Pull request merged**
